src/websocket/server.cc

```diff
@@ -166,34 +200,208 @@ future<> connection::read_http_upgrade_request() {
+            // FIXME: implement error handling
+            switch(_websocket_parser.opcode()) {
+                // We do not distinguish between these 3 types.
+                case opcodes::CONTINUATION:
+                case opcodes::TEXT:
+                case opcodes::BINARY:
+                    return _input_buffer.push_eventually(_websocket_parser.result());
+                case opcodes::CLOSE:
+                    wlogger.debug("Received close frame.");
+                    /*
+                     * datatracker.ietf.org/doc/html/rfc6455#section-5.5.1
+                     */
+                    return close(true);
+                case opcodes::PING:
+                    return handle_ping();
+                    wlogger.debug("Received ping frame.");
+                case opcodes::PONG:
+                    wlogger.debug("Received pong frame.");
+                    return handle_pong();
+                default:
+                    // Invalid - do nothing.
+                    ;
+            }
+        } else if (_websocket_parser.eof()) {
+            return close(false);
         }
-        //FIXME: implement
-        wlogger.info("Received: {}", buf.get());
+        wlogger.debug("Reading from socket has failed.");
+        return close(true);
     });
 }
 
 future<> connection::read_loop() {
     return read_http_upgrade_request().then([this] {
-        return do_until([this] {return _done;}, [this] {
-            return read_one();
+        return when_all(_handler(_input, _output),
+            do_until([this] {return _done;}, [this] {return read_one();})
+        ).then([] (std::tuple<future<>, future<>> joined) {
+            try {
+                std::get<0>(joined).get();
+            } catch (...) {
+                wlogger.debug("Handler exception encountered: {}", 
+                        std::current_exception());
+            }
+            try {
+                std::get<1>(joined).get();
+            } catch (...) {
+                wlogger.debug("Read exception encountered: {}", 
+                        std::current_exception());
+            }
+            return make_ready_future<>();
+        }).finally([this] {
+            return _read_buf.close();
         });
-    }).then_wrapped([this] (future<> f) {
-        if (f.failed()) {
-            wlogger.error("Read failed: {}", f.get_exception());
+    });
+}
+
+future<> connection::close(bool send_close) {
+    return [this, send_close]() {
+        if (send_close) {
+            return send_data(opcodes::CLOSE, temporary_buffer<char>(0));
+        } else {
+            return make_ready_future<>();
         }
-        return _replies.push_eventually({});
-    }).finally([this] {
-        return _read_buf.close();
+    }().then([this] {
+        _done = true;
+        return when_all(_input.close(), _output.close()).discard_result();
     });
 }
 
+future<> connection::send_data(opcodes opcode, temporary_buffer<char>&& buff) {
+    sstring data;
```

Comment:
@nyh regarding using string::append to generate the header, I think that with the new implementation based on scattered_message it's actually quite optimal. Under the assumption that small string opitimization is in place (and it is both for gcc and clang) and generally allows storing ~15 bytes in-place without any allocation, this header will be constructed entirely on the stack and later moved straight to the scattered message, because it's never longer than 15 bytes.